src/dune_rules/jsoo/jsoo_rules.ml

```diff
@@ -41,136 +41,223 @@ let js_env = compute_env ~mode:JS
+      | maybe_effects :: rest when String.is_prefix maybe_effects ~prefix:"--effects=" ->
+        let backend =
+          Option.bind
+            (String.drop_prefix maybe_effects ~prefix:"--effects=")
+            ~f:effects_of_string
+        in
+        (match backend with
+         | Some backend -> loop { acc with effects = Some backend } rest
+         | None -> loop acc rest)
       | _ :: rest -> loop acc rest
     in
     loop default l
   ;;
 
-  let to_flags t =
-    List.concat_map (get t) ~f:(function
-      | "toplevel", true -> [ "--toplevel" ]
-      | "toplevel", false -> []
-      | name, true -> [ "--enable"; name ]
-      | name, false -> [ "--disable"; name ])
+  let backward_compatible_effects ~jsoo_version str =
+    match str with
+    | None ->
+      (* For jsoo, this means unsupported effects. For wasmoo, this means effects go
+         through the Javascript Promise API. *)
+      None
+    | Some Cps ->
+      let v6_or_higher =
+        match jsoo_version with
+        | Some v ->
+          (match Version.compare v (6, 0) with
+           | Gt | Eq -> true
+           | Lt -> false)
+        | None -> false
+      in
+      if v6_or_higher then Some "--effects=cps" else Some "--enable=effects"
+    | Some Double_translation ->
+      (* For js_of_ocaml < 6.0, this flag does not exist and will raise an error,
+         which is fine. *)
+      Some "--effects=double-translation"
   ;;
-end
-
-module Version = struct
-  type t = int * int
 
-  let of_string s : t option =
-    let s =
-      match
-        String.findi s ~f:(function
-          | '+' | '-' | '~' -> true
-          | _ -> false)
-      with
-      | None -> s
-      | Some i -> String.take s i
-    in
-    try
-      match String.split s ~on:'.' with
-      | [] -> None
-      | [ major ] -> Some (int_of_string major, 0)
-      | major :: minor :: _ -> Some (int_of_string major, int_of_string minor)
-    with
-    | _ -> None
-  ;;
-
-  let compare (ma1, mi1) (ma2, mi2) =
-    match Int.compare ma1 ma2 with
-    | Eq -> Int.compare mi1 mi2
-    | n -> n
-  ;;
-
-  let impl_version bin =
-    let* _ = Build_system.build_file bin in
-    Memo.of_reproducible_fiber
-    @@ Process.run_capture_line ~display:Quiet Strict bin [ "--version" ]
-    |> Memo.map ~f:of_string
+  let to_flags ~jsoo_version t =
+    List.filter_opt
+      [ (match t.toplevel with
+         | Some true -> Some "--toplevel"
+         | _ -> None)
+      ; backward_compatible_effects ~jsoo_version t.effects
+      ; (match t.js_string with
+         | Some true -> Some "--enable=use-js-string"
+         | Some false -> Some "--disable=use-js-string"
+         | None -> None)
+      ]
   ;;
 
-  let version_memo = Memo.create "jsoo-version" ~input:(module Path) impl_version
-
-  let jsoo_version jsoo =
-    match jsoo with
-    | Ok jsoo_path -> Memo.exec version_memo jsoo_path
-    | Error e -> Action.Prog.Not_found.raise e
+  let remove_effect_flags flags =
```

Comment:
It feels weird to special case the effects flags only.
Could we instead remove all flags already covered by the config ?

Reply:
Done. However, it looks like this step of removing some flags has the unintended side effects of removing user-added flags as well. I added a test that is supposed to fail due to incompatible `--effects` flags, and the `--effects=double-translation` flag in the library's dune file is simply discarded.

Reply:
I thinks that's the point: we want to use the flags from where the library is used.
Some default flags are often sets at the root of the dune workspace. It would be cumbersome if we had to remove them for libraries.

Reply:
Oh I see, I’ve been looking at this the wrong way.